File: 11_ContainerWithMostWater.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int l = 0;
        int r = height.size()-1;
        int maxleft = height[l];
        int maxright = height[r];
        int maxvolume = 0;
        while(l<r){
            if (maxleft >= maxright){
                maxvolume = max(maxvolume,(r-l)*maxright);
                r--;
                maxright = max(maxright, height[r]);
            }
            if (maxleft < maxright) {
                maxvolume = max(maxvolume,(r-l)*maxleft);
                l++;
                maxleft = max(maxleft, height[l]);
            }
        }
    
    return maxvolume;
    }
};

File: 11_ContainerWithMostWater.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        l = 0
        r = len(height)-1
        maxheightleft = height[l]
        maxheightright = height[r]
        maxv = 0
        while(l<r):
            if maxheightleft >= maxheightright:
                maxv = max(maxv, maxheightright*(r-l))
                r-=1
                maxheightright=max(maxheightright, height[r])
            if maxheightleft < maxheightright:
                maxv = max(maxv, maxheightleft*(r-l))
                l+=1
                maxheightleft=max(maxheightleft, height[l])
        return maxv